src/omv/modules/py_tf.c

```diff
@@ -803,31 +708,384 @@ STATIC void py_tf_regression_input_callback(void *callback_data,
     }
 }
 
-STATIC mp_obj_t py_tf_regression(mp_obj_t model_obj, mp_obj_t array_obj) {
+typedef struct py_tf_model_output_obj {
+    mp_obj_base_t base;
+    rectangle_t *roi;
+    void *model_output;
+    libtf_parameters_t *params;
+    // Pre-compute for lookup speed.
+    size_t output_size;
+    mp_obj_t x, y, w, h, rect;
+    mp_obj_t input_height, input_width, input_channels;
+    mp_obj_t output_height, output_width, output_channels;
```

Comment:
I don't mind allocating a tuple internally, but, the result to the end user has to be static. We cannot allocate tuples dynamically without a huge speed hit.

Are you thinking:

result.input_shape[0] versus result.input_height, result.input_shape[1] versus result.input_width, and result.input_shape[2] versus result.input_channels?

Reply:
In `py_tf_invoke_output_callback` you already allocate tuples for the `rect`. There won't be an extra alloc when you access that tuple in `py_tf_model_output_attr` via `self->input_shape`.

Btw why is there a `rect` and `x, y, w, h` it seems redundant.

Reply:
Also it seems that all this information is already accessible from the model object, right ? If that's the case, why duplicate it in the output object and waste time copying it all over ? The callback should have access to the model object in Python. They still use functions but we can switch them to attributes.

Reply:
I changed everything to shape into the model output object but didn't switch to passing the main object in the callback.  I also removed x/y/w/h and just have the rect object. Since we released the API previously as functions it will break any code in the wild to turn them into attributes. The "()" at the end of any previous function call that gets turned into an attribute will cause MicroPython to then try to call the value of the attribute returned.

I agree we should use attributes moving forward. But, I don't think we should go break the API of older stuff.

Reply:
Example code updated.

Reply:
> I agree we should use attributes moving forward. But, I don't think we should go break the API of older stuff.

Please implement what we discussed in the meeting. The API to access those attributes isn't that common, before the callback you had no reason to read the input/output shape or channels. Also APIs change, we don't have to get stuck with a poor API. The change is for the better.

Reply:
Done